scheduler/executor/workflow_executor.go

```diff
@@ -0,0 +1,431 @@
+				return
+			} else {
+				w.snapshot.UpdateLastExecutionTime(nameOfSchedule, scheduleTime)
+			}
+		}
+
+		// Register or deregister the schedule from the scheduler
+		if !*s.Active {
+			w.goCronInterface.DeRegister(ctx, s)
+		} else {
+			err := w.goCronInterface.Register(ctx, s, funcRef)
+			if err != nil {
+				w.metrics.ScheduleRegistrationFailure.Inc()
+				logger.Errorf(ctx, "unable to register the schedule %+v due to %v", s, err)
+			}
+		}
+	} // Done iterating over all the read schedules
+	return nil
+}
+
+func (w *workflowExecutor) RunCatchupSystem(ctx context.Context) chan error {
+	catchupCtx, cancelCatchUp := context.WithCancel(ctx)
+	catchUpContextWithLabel := contextutils.WithGoroutineLabel(catchupCtx, catchupRoutineLabel)
+	catchUpErrChan := make(chan error, 1)
+	go func(ctx context.Context) {
+		pprof.SetGoroutineLabels(ctx)
+		defer func() {
+			if err := recover(); err != nil {
+				w.metrics.CatchupPanicCounter.Inc()
+				logger.Fatalf(ctx, fmt.Sprintf("caught panic: %v [%+v]", err, string(debug.Stack())))
+			}
+		}()
+		defer cancelCatchUp()
+		catchUpErrChan <- w.ReadAndCatchupSchedules(ctx)
+	}(catchUpContextWithLabel)
+
+	return catchUpErrChan
+}
+
+func (w *workflowExecutor) RunCheckpointerSystem(ctx context.Context) {
+	checkPointerCtx, checkPointerCancel := context.WithCancel(ctx)
+	checkPointerCtxWithLabel := contextutils.WithGoroutineLabel(checkPointerCtx, checkPointerRoutineLabel)
+	go func(ctx context.Context) {
+		pprof.SetGoroutineLabels(ctx)
+		defer func() {
+			if err := recover(); err != nil {
+				w.metrics.CheckPointPanicCounter.Inc()
+				logger.Fatalf(ctx, fmt.Sprintf("caught panic: %v [%+v]", err, string(debug.Stack())))
+			}
+		}()
+		defer checkPointerCancel()
+		wait.UntilWithContext(ctx, w.CheckPointState, snapshotWriterSleepTime * time.Second)
+	}(checkPointerCtxWithLabel)
+}
+
+func (w *workflowExecutor) RunScheduleRegistrationSystem(ctx context.Context) {
+	scheduleRegisterCtx, scheduleRegisterCtxCancel := context.WithCancel(ctx)
+	scheduleRegisterCtxWithLabel := contextutils.WithGoroutineLabel(scheduleRegisterCtx, scheduleRegisterRoutineLabel)
+	go func(ctx context.Context) {
+		pprof.SetGoroutineLabels(ctx)
+		defer func() {
+			if err := recover(); err != nil {
+				w.metrics.CheckPointPanicCounter.Inc()
+				logger.Fatalf(ctx, fmt.Sprintf("caught panic: %v [%+v]", err, string(debug.Stack())))
+			}
+		}()
+		defer scheduleRegisterCtxCancel()
+		wait.UntilWithContext(ctx, func(ctx context.Context){
+			err := w.ReadAndRegisterSchedules(ctx)
+			if err != nil {
+				logger.Errorf(ctx, "Failed to iterated over all schedules in the current run due to %v", err)
+			}
+		}, registrationSleepTime * time.Second)
+	}(scheduleRegisterCtxWithLabel)
+}
+
+func (w *workflowExecutor) Run(ctx context.Context) error {
+	// Start the go routine to catchup all the schedules
+	// Create catchup function label to be used for creating the child context
+	catchUpErrChan := w.RunCatchupSystem(ctx)
```

Comment:
You should create the cancel func here and defer it so that in case any other goroutine returns you can cancel others